exonum-java-binding-common/src/test/java/com/exonum/binding/common/proofs/list/ListProofStructureValidatorTest.java

```diff
@@ -0,0 +1,239 @@
+
+    validator = createListProofStructureValidator();
+    validator.visit(root);
+
+    assertThat(validator.check(), is(ListProofStatus.VALID));
+  }
+
+  @Test
+  void visit_ProofRightValue() {
+    ListProof left = new ListProofHashNode(H1);
+    ListProof right = leafOf(V2);
+    ListProofBranch root = new ListProofBranch(left, right);
+
+    validator = createListProofStructureValidator();
+    validator.visit(root);
+
+    assertThat(validator.check(), is(ListProofStatus.VALID));
+  }
+
+  @Test
+  void visit_InvalidTreeHasNoElements() {
+    ListProof left = new ListProofHashNode(H1);
+    ListProofBranch root = new ListProofBranch(left, null);
+
+    validator = createListProofStructureValidator();
+    validator.visit(root);
+
+    assertThat(validator.check(), is(ListProofStatus.INVALID_TREE_NO_ELEMENTS));
+  }
+
+  @Test
+  void visit_UnbalancedInTheRightSubTree() {
+    ListProofBranch root = new ListProofBranch(
+        new ListProofBranch(leafOf(V1),
+            new ListProofHashNode(H2)),
+        leafOf(V3) // <-- A value at the wrong depth.
+    );
+
+    validator = createListProofStructureValidator();
+    validator.visit(root);
+
+    assertThat(validator.check(), is(ListProofStatus.INVALID_NODE_DEPTH));
+  }
+
+  @Test
+  void visit_UnbalancedInTheLeftSubTree() {
+    ListProofBranch root = new ListProofBranch(
+        leafOf(V1), // <-- A value at the wrong depth.
+        new ListProofBranch(leafOf(V2),
+            new ListProofHashNode(H3))
+    );
+
+    validator = createListProofStructureValidator();
+    validator.visit(root);
+
+    assertThat(validator.check(), is(ListProofStatus.INVALID_NODE_DEPTH));
+  }
+
+  @Test
+  void visit_MaxAllowedDepth() {
+    int depth = ListProofStructureValidator.MAX_NODE_DEPTH;
+    ListProof root = generateRightLeaningProofTree(depth, leafOf(V1));
+
+    validator = createListProofStructureValidator();
+    root.accept(validator);
+
+    assertThat(validator.check(), is(ListProofStatus.VALID));
+  }
+
+  @Test
+  void visit_UnbalancedElementNodeTooDeep() {
+    int depth = ListProofStructureValidator.MAX_NODE_DEPTH + 1;
+    ListProof root = generateRightLeaningProofTree(depth, leafOf(V1));
+
+    validator = createListProofStructureValidator();
+    root.accept(validator);
+
+    assertThat(validator.check(), is(ListProofStatus.INVALID_ELEMENT_NODE_DEPTH));
+  }
+
+  @Test
+  void visit_UnbalancedHashNodeTooDeep() {
+    int depth = ListProofStructureValidator.MAX_NODE_DEPTH + 1;
+    ListProof root = generateRightLeaningProofTree(depth, new ListProofHashNode(H2));
+
+    validator = createListProofStructureValidator();
+    root.accept(validator);
+
+    assertThat(validator.check(), is(ListProofStatus.INVALID_HASH_NODE_DEPTH));
+  }
+
+  @Test
+  void visit_UnbalancedHashNodesOnlyLeafs() {
+    ListProofBranch root = new ListProofBranch(
+        new ListProofBranch(
+            leafOf(V1),
+            new ListProofHashNode(H1) // <-- left leaf is hash node
```

Comment:
The comment is not correct (it's a right node, and this branch is OK, so I'd leave it out).